### DISCUSSION TITLE
Form reloadDocument for resource route download with success/error info

### DISCUSSION THREAD

**Discussion by humphd** [remix-run/remix#4964] [Q&A] [answered]
I'm trying to find a good solution for dealing with success/error info after loading a resource route that returns a file for download.  It seems I have two options, and I want a hybrid, somewhere between the two.

For context, imagine a page with a dialog/form, where you fill out some preferences and then use those values to dynamically generate a file that the client wants to download.  So you click a **Download** button, and it a resource route:

```ts
// Resource route at /resource
export const loader: LoaderFunction = async () => {
  return new Response('Hello World!', {
    status: 200,
    headers: {
      'Content-Disposition': 'attachment; filename="result.txt"',
      'Content-Type': 'text/whatever',
    },
  });
};
```

I seem to have two ways to deal with this flow:

1. Letting the browser do it all without JS using `<Form action="/resource" reloadDocument>...<button type="submit">Download</button></Form>`
2. Do everything in JS using a mix of `fetch()`, a `Blob`,  and `URL.createObjectURL()` so that the browser can download the data.

The first option seems more in the spirit of Remix, but I really need to know:

1. When the request has finished
2. What the `status` was (did it work or fail?) so I can update the UI, show any error state, offer to try again, etc.

It doesn't appear that `<Form reloadDocument>` is compatible with transitions or any sort of state via JS (see other similar questions: https://github.com/remix-run/remix/discussions/4851, https://github.com/remix-run/remix/discussions/4698).  This makes sense, but is pretty unsatisfying.

Am I missing another option, or should I just go around Remix and do everything client side?

**Comment by machour** [comment, marked as answer]
reloadDocument is really what it says: it performs a full page reload, so JavaScript is out of the picture.

If you want a document, that the browser would natively download to your disk, to give back a download status, then this definitely need to happen on the client side, with JavaScript. You could maybe leverage useFetcher() but it won't report download progress as far as I know. So your best bet is to handle this yourself like you mentioned. 

On a side note, this isn't really going around remix. It's going around the browser :) it's totally fine and acceptable if you keep this in mind: One little danger is that your JavaScript may break for whatever reason, and the download wouldn't happen. So I would make sure to provide an escape hatch with a discrete link for direct download in case things go south. And that's part of what Remix preaches 🙌

**Reply by humphd** [reply]
Thanks for this @machour.  It sounds like there isn't something more that I'm missing, which is good to know.  I appreciate the answer as I explore the edges of Remix.

**Comment by justinwaite** [comment]
I've run into a similar problem. It would be nice to be able to use the Remix `Link` component to link to the resource route that generates the pdf, but this doesn't work. I have to use a plain anchor tag `a` to get the file to download, but then I don't have any navigation state information (is the resource route still loading?).
